src/components/Users/Users.js

```diff
@@ -1,12 +1,49 @@
-import User from '../User/User';
-import './Users.css';
+import User from "../User/User";
+import "./Users.css";
+import { useState } from "react";
 
 const Users = ({ users = [] }) => {
+  const [expand, setExpand] = useState([]);
+
+  const handleExpand = (id) => {
+    if (!expand.includes(id)) {
+      const newExpand = [...expand, id];
+      setExpand(newExpand);
+    } else {
+      const filteredExpand = expand.filter((currentId) => {
+        return currentId !== id;
+      });
+      setExpand(filteredExpand);
+    }
+  };
+
+  const expandAllHandler = () => {
+    const expandAll = users.map((user) => {
+      return user.id;
+    });
+    setExpand(expandAll);
+  };
+
+  const collapseAllHandler = () => {
+    setExpand([]);
+  };
+
   return (
     <article className="Users">
+      <button onClick={expandAllHandler}>Expand All</button>
+      <button onClick={collapseAllHandler}>Collapse All</button>
       {users.map((user) => {
         const { id } = user;
-        return <User key={id} user={user} />;
+        return (
+          <User
+            key={id}
+            user={user}
+            expand={expand}
```

Comment:
I think it is better to pass a boolean prop for `expand` here instead of passing the entire array. In other words, I would prefer to check `expand.includes(id)` here rather than in the `<User />` component. There are a couple of reasons why:

1. The principle of lifting state up. `<Users />` should have complete control over whether a `<User />` is expanded. Complete control means complete responsibility. So `<Users />` should be responsible for checking if `expanded` includes the user's id.
2. General software design principles. In general we want every part of our system to have one clearly defined job that they completely control. This makes it easier to debug, refactor, and extend the functionality of the software. Related to this, each part of the system should only "know" the data that they need to do their specific job. In this app, that means that `<Users />` should have the array of ids because it needs to know that in order to tell each child `<User />` to expand or not expand. A single `<Users />` component should not have the array of ids because it doesn't need to know about other users in order to do its job.

Reply:
Thank You.